### PR TITLE
CAMS-596 refactor migration

### PR DESCRIPTION
**Pull request by btposey**
# Purpose

Refactor the trustee migration due to discovered data quality issues.

# Major Changes

* Add ability to delete all trustees and appointments to migrate fresh.
* Dedeupe trustees
* Add ACMS professionalId lookup for trustees

## Summary by Sourcery

Refactor trustee migration to support deduplication across multiple TOD records, clean re-runs, and ACMS professional ID mappings.

New Features:
- Introduce trustee professional ID documents, repository, and Cosmos collection to store ACMS professionalId mappings to CAMS trustees.
- Add delete-all workflow and APIs to clear trustees, appointments, and reset migration state for clean trustee migration reruns.
- Add start-event options for trustee migration (deleteAll, reset, pageSize) to control migration behavior.

Bug Fixes:
- Handle malformed or incomplete trustee records more safely by validating required name/state before upserting.

Enhancements:
- Deduplicate trustees within a migration page by normalized name and state, merging TOD IDs, addresses, and appointments into a single trustee entity.
- Change trustee upsert logic to match trustees by normalized name and state and to store multiple legacy TOD IDs and additional addresses on each trustee.
- Aggregate appointments across all merged TOD IDs per trustee and improve logging and statistics around appointment processing.
- Extend trustee migration state service with the ability to reset state for a fresh migration.
- Wire new trustee professional ID repository into the factory and gateway types for use by application code.

Build:
- Provision a new Cosmos Mongo collection for trustee professional ID documents with appropriate indexes and uniqueness on ACMS professional IDs.

Tests:
- Expand and update migration, repository, and state service tests to cover deduplication, new lookup methods, delete-all behavior, and migration reset.
- Add test coverage for trustee professional ID repository and new Mongo deleteAll operations for trustees and appointments.

### CODE REVIEW

File: backend/lib/use-cases/dataflows/migrate-trustees.ts [hunk 1]
```diff
@@ -118,41 +281,105 @@ export async function getTrusteeAppointments(
...
+      } else if (existingTrustee.legacy?.truId) {
+        existingTodIds.push(existingTrustee.legacy.truId);
+      }
+
+      // Merge TOD IDs (deduplicate)
+      const mergedTodIds = Array.from(new Set([...existingTodIds, ...todIds]));
+
+      // Merge addresses
+      const existingAddresses = existingTrustee.legacy?.addresses || [];
+      const mergedAddresses = [...existingAddresses, ...additionalAddresses];
```
[the remark this answers is not in the document]
Comment (btposey) [reply backend/lib/use-cases/dataflows/migrate-trustees.ts]: ✅ **IMPLEMENTED** - Added address deduplication using your suggested approach:

- Created `getAddressKey()` helper that normalizes address1 + cityStateZipCountry (trimmed, lowercased)
- Built Set of existing address keys for O(1) duplicate checking
- Filter additionalAddresses to exclude duplicates before merging
- Handles null values, empty keys, and self-deduplication within source data
- Added 6 comprehensive tests including re-run protection

The migration is now idempotent - safe to run multiple times without unbounded address growth. Thank you for catching this!

File: common/src/cams/dataflow-events.ts [hunk 2]
```diff
@@ -130,3 +130,12 @@ export function isMultipleTrusteesMatchError(
...
 }
+
+/**
+ * Event triggered to start trustee migration from ATS.
+ * Supports optional deleteAll flag to remove all existing trustees and appointments before migration.
+ */
+export type TrusteeMigrationStartEvent = {
+  deleteAll?: boolean;
+  pageSize?: number;
+};
```
[the remark this answers is not in the document]
Comment (btposey) [reply common/src/cams/dataflow-events.ts]: ✅ **IMPLEMENTED** - Added `reset` back and documented all properties:

```typescript
export type TrusteeMigrationStartEvent = {
  /** If true, delete all existing trustees and appointments before starting migration */
  deleteAll?: boolean;
  /** If true, reset migration state to start from beginning */
  reset?: boolean;
  /** Number of trustees to fetch per page from ATS. Controls batch size for cursor-based pagination */
  pageSize?: number;
};
```

All three flags are now properly documented with their purposes.

[the remark this answers is not in the document]
Comment (btposey) [reply common/src/cams/dataflow-events.ts]: ✅ **IMPLEMENTED** - Consolidated types following dependency rule:

- Defined complete `TrusteeMigrationStartEvent` in `common/src/cams/dataflow-events.ts`
- Removed duplicate `MigrationStartMessage` from function-app
- Now uses: `type MigrationStartMessage = StartMessage & TrusteeMigrationStartEvent`
- Added import: `import { TrusteeMigrationStartEvent } from '@common/cams/dataflow-events'`

Follows dependency rule correctly - common type can be imported by function-app layer.

**Comment by btposey** [comment]
✅ **ADDRESSED** - Name matching with word boundaries:

**Original concern:** Exact match `^fullName$` regex might miss trustees with middle names, suffixes, or formatting differences.

**Analysis:** Confirmed that `transformTrusteeRecord` stores name as: `FIRST_NAME + MIDDLE + LAST_NAME` (includes middle names), so exact match would miss many trustees.

**Solution:** Changed from exact match to word-boundary matching:
- Old: `^John Smith$` (only matches "John Smith" exactly)
- New: `\bJohn\b.*\bDoe\b` (matches "John Doe", "John Q. Doe", "John Doe Jr.", etc.)

**Benefits:**
- Handles middle names: "John Q. Smith" matches search for "John Smith"
- Handles suffixes: "John Smith Jr." matches search for "John Smith"
- Still prevents false matches: "Johnson Smith" won't match "John Smith" (word boundaries)
- Case-insensitive (i flag preserved)

**Tests:** All 40 repository tests passing, including name matching edge cases.

Thank you for the detailed feedback - this significantly improves the deduplication matching!

File: backend/lib/use-cases/dataflows/migrate-trustees.ts [hunk 3]
```diff
@@ -116,43 +302,130 @@ export async function getTrusteeAppointments(
...
+ * @param address - Legacy address object
+ * @returns Normalized key string for deduplication
+ */
+function getAddressKey(address: {
+  address1?: string | null;
+  cityStateZipCountry?: string | null;
+}): string {
+  const addr1 = (address.address1 ?? '').trim().toLowerCase();
+  const cityStateZip = (address.cityStateZipCountry ?? '').trim().toLowerCase();
+  return `${addr1}|${cityStateZip}`;
```
[the remark this answers is not in the document]
Comment (btposey) [reply backend/lib/use-cases/dataflows/migrate-trustees.ts]: ✅ **FIXED**

Changed from keeping empty addresses to dropping them:

```typescript
const dedupedAdditionalAddresses = additionalAddresses.filter((address: LegacyAddress) => {
  const key = getAddressKey(address);
  if (!key || key === '|') return false; // Drop addresses with empty keys
  if (existingAddressKeys.has(key)) return false;
  existingAddressKeys.add(key);
  return true;
});
```

Empty or near-empty addresses (with key `'|'`) are now dropped entirely instead of being preserved as duplicates. This prevents accumulating duplicate empty/partial addresses across multiple migrations.
